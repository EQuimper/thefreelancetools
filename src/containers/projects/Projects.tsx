import { Button } from 'evergreen-ui';
import { observer } from 'mobx-react';
import * as React from 'react';

import { store } from '@freelance-tool/models';
import { ModalTypeEnum } from '@freelance-tool/types';
// import styled from 'styled-components';

// const Root = styled.div`
//   display: grid;
// `;

@observer
class Projects extends React.Component {
  state = {};
  _openNewProjectModal = () => {
    store.modalsManager.open({
      modalTitle: 'New Project',
      modalType: ModalTypeEnum.NEW_PROJECT,
      params: null,
    });
  }

  render() {
    const { projects } = store;
    return (
      <div>
        Hello world!!
        <Button onClick={this._openNewProjectModal}>Create New Project</Button>
        <div>
          {projects.projects.map(el => (
            <div key={String(el.id)}>
              <p>{el.name}</p>
            </div>
          ))}
        </div>
      </div>
    );
  }
}

export default Projects;
