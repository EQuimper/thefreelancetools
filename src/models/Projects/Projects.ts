import { toaster } from 'evergreen-ui';
import { destroy, getParent, types } from 'mobx-state-tree';
import * as uuid from 'uuid/v4';

import { ProjectPriorityEnum } from '@freelance-tool/types';

interface ElapsedTime {
  hours: number;
  minutes: number;
  seconds: number;
  totalSeconds: number;
}

export const Task = types
  .model('Task', {
    id: types.optional(types.identifier(), () => uuid()),
    elapsedTime: types.model({
      hours: 0,
      minutes: 0,
      seconds: 0,
      totalSeconds: 0,
    }),
    startAt: types.maybe(types.Date),
    endAt: types.maybe(types.Date),
    name: types.string,
  })
  .actions(self => ({
    end(elapsedTime: ElapsedTime) {
      self.endAt = new Date();

      self.elapsedTime = {
        hours: elapsedTime.hours,
        minutes: elapsedTime.minutes,
        seconds: elapsedTime.seconds,
        totalSeconds: elapsedTime.totalSeconds,
      };
    },
  }))
  .actions(self => ({
    afterCreate() {
      self.startAt = new Date();
    },
  }));

export const Project = types
  .model('Project', {
    id: types.optional(types.identifier(), () => uuid()),
    name: types.string,
    description: types.string,
    priority: types.enumeration('priority', ['HIGH', 'MEDIUM', 'LOW']),
    tasks: types.optional(types.array(Task), []),
  })
  .actions(self => ({
    remove() {
      getParent(self, 2).remove(self);
    },
  }))
  .actions(self => ({
    updateName(name: string) {
      self.name = name;
    },
  }))
  .actions(self => ({
    updateDescription(description: string) {
      self.description = description;
    },
  }))
  .actions(self => ({
    updatePriority(priority: ProjectPriorityEnum) {
      self.priority = priority;
    },
  }))
  .actions(self => ({
    addTask(task: typeof Task.Type) {
      self.tasks.push(task);
    },
  }))
  .actions(() => ({
    afterCreate() {
      toaster.success('Project successfully created!');
    },
  }))
  .views(self => ({
    get totalTime(): ElapsedTime {
      return self.tasks.reduce(
        (acc, currentValue) => {
          return {
            hours: currentValue.elapsedTime.hours + acc.hours,
            minutes: currentValue.elapsedTime.minutes + acc.minutes,
            seconds: currentValue.elapsedTime.seconds + acc.seconds,
            totalSeconds:
              currentValue.elapsedTime.totalSeconds + acc.totalSeconds,
          };
        },
        {
          hours: 0,
          minutes: 0,
          seconds: 0,
          totalSeconds: 0,
        },
      );
    },
  }));

export const Projects = types
  .model('Projects', {
    projects: types.optional(types.map(Project), {}),
  })
  .actions(self => ({
    addProject(project: typeof Project.Type) {
      self.projects.put(project);
    },
  }))
  .actions(self => ({
    remove(project: typeof Project.Type) {
      destroy(project);
    },
  }))
  .views(self => ({
    get totalProjectsAmount() {
      return self.projects.size;
    },
  }))
  .views(self => ({
    getProjectById(id: string) {
      return self.projects.get(id);
    },
  }))
  .views(self => ({
    get getProjects() {
      return self.projects;
    },
  }))
  .views(self => ({
    get getProjectsOptions(): Array<{ label: string; value: string }> {
      const options: Array<{ label: string; value: string }> = [];

      self.projects.forEach((val, key) => {
        options.push({
          value: key,
          label: val.name,
        });
      });

      return options;
    },
  }));
