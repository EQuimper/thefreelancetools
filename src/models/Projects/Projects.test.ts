import { onPatch, onSnapshot } from 'mobx-state-tree';

import { Project, Projects } from './Projects';

describe('Project model', () => {
  it('should create a instance of a model', () => {
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });

    expect(myProject.name).toBe(name);
    expect(myProject.description).toBe(description);
  });

  it('should be able to update the name', () => {
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });
    const updatedName = 'Hello World';
    const patches: any[] = [];

    onPatch(myProject, patch => {
      patches.push(patch);
    });

    expect(myProject.name).toBe(name);

    myProject.updateName(updatedName);

    expect(myProject.name).toBe(updatedName);

    expect(patches).toMatchSnapshot();
  });

  it('should be able to update the description', () => {
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });
    const updatedDescription = 'Hello World';
    const patches: any[] = [];

    onPatch(myProject, patch => {
      patches.push(patch);
    });

    expect(myProject.description).toBe(description);

    myProject.updateDescription(updatedDescription);

    expect(myProject.description).toBe(updatedDescription);

    expect(patches).toMatchSnapshot();
  });
});

describe('Projects model', () => {
  it('should create a instance of a model', () => {
    const projects = Projects.create();

    expect(projects.projects.length).toBe(0);
  });

  it('should be able to add new project when create', () => {
    const projects = Projects.create();
    const states: any[] = [];
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });
    const name2 = 'My Second Project';
    const myProject2 = Project.create({ name: name2, id: '456', description });

    onSnapshot(projects, snapshot => {
      states.push(snapshot);
    });

    projects.addProject(myProject);

    expect(projects.projects.length).toBe(1);
    expect(projects.projects[0].name).toBe(name);

    projects.addProject(myProject2);

    expect(projects.projects.length).toBe(2);
    expect(projects.projects[1].name).toBe(name2);

    expect(states).toMatchSnapshot();
  });

  it('should be able to get a computed value of the amount of projects', () => {
    const projects = Projects.create();
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });
    const name2 = 'My Second Project';
    const myProject2 = Project.create({ name: name2, id: '456', description });

    projects.addProject(myProject);

    expect(projects.totalProjectsAmount).toBe(1);

    projects.addProject(myProject2);

    expect(projects.totalProjectsAmount).toBe(2);
  });

  it('should be able to remove a project', () => {
    const projects = Projects.create();
    const states: any[] = [];
    const name = 'My Project';
    const description = 'This is a description';
    const myProject = Project.create({ name, id: '123', description });
    const name2 = 'My Second Project';
    const myProject2 = Project.create({ name: name2, id: '456', description });

    onSnapshot(projects, snapshot => {
      states.push(snapshot);
    });

    projects.addProject(myProject);
    projects.addProject(myProject2);

    expect(projects.projects.length).toBe(2);

    myProject.remove();

    expect(projects.projects.length).toBe(1);

    expect(states).toMatchSnapshot();
  });
});
