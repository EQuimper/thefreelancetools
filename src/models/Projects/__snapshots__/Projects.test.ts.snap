// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project model should be able to update the description 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/description",
    "value": "Hello World",
  },
]
`;

exports[`Project model should be able to update the name 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/name",
    "value": "Hello World",
  },
]
`;

exports[`Project model should be able to update the priority 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/priority",
    "value": "HIGH",
  },
]
`;

exports[`Projects model should be able to add new project when create 1`] = `
Array [
  Object {
    "projects": Array [
      Object {
        "description": "This is a description",
        "id": "123",
        "name": "My Project",
        "priority": "LOW",
      },
    ],
  },
  Object {
    "projects": Array [
      Object {
        "description": "This is a description",
        "id": "123",
        "name": "My Project",
        "priority": "LOW",
      },
      Object {
        "description": "This is a description",
        "id": "456",
        "name": "My Second Project",
        "priority": "LOW",
      },
    ],
  },
]
`;

exports[`Projects model should be able to remove a project 1`] = `
Array [
  Object {
    "projects": Array [
      Object {
        "description": "This is a description",
        "id": "123",
        "name": "My Project",
        "priority": "LOW",
      },
    ],
  },
  Object {
    "projects": Array [
      Object {
        "description": "This is a description",
        "id": "123",
        "name": "My Project",
        "priority": "LOW",
      },
      Object {
        "description": "This is a description",
        "id": "456",
        "name": "My Second Project",
        "priority": "LOW",
      },
    ],
  },
  Object {
    "projects": Array [
      Object {
        "description": "This is a description",
        "id": "456",
        "name": "My Second Project",
        "priority": "LOW",
      },
    ],
  },
]
`;
